### DISCUSSION TITLE
Blockchain & Oqtane Security

### DISCUSSION THREAD

**Discussion by Sharafudeen** [oqtane/oqtane.framework#3222] [General] [answered]
Hi,

I’m thinking how to secure all of my Oqtane modules. What is the recommended way to implement blockchain in Oqtane for security? I wanted to encrypt & decrypt (like in ASPNET Core “Data Protection”) all of my data transactions (not only payment transactions).

Is that correct way of thinking to implement blockchain as a “Middleware or Custom Oqtane Module or something else”?

Examples:
[Blockchain Sample 1](https://www.c-sharpcorner.com/article/write-your-first-smart-contract-on-stratis-blockchain/) 

[Blockchain Framework - BlockCore](https://github.com/block-core/blockcore)

**Comment by leigh-pointer** [comment]
@Sharafudeen on the face of it by the examples I dont see you having any problems.

**Reply by Sharafudeen** [reply]
I mean that, what is the recommended way to implement security using blockchain in Oqtane?

**Reply by sbwalker** [reply]
@Sharafudeen Blockchain is generally used for very specific types of applications or use cases, and since Oqtane is intended to be a general purpose framework it does not use blockchain concepts. I have not seen any examples of modules using blockchain in Oqtane. And in regards to your example of encryption/decryption there are plenty of ways to accomplish this even without using blockchain. So basically I am suggesting that you do your research and try to choose the right technology for the application - don't just choose blockchain because it's popular. Either way, please share your experiences with the Oqtane community so that others can benefit from it.

**Reply by Sharafudeen** [reply]
@sbwalker & @leigh-pointer Thanks. I would like to implement followings in Oqtane. How can I do that?

[10 Best Practices](https://www.syncfusion.com/blogs/post/10-practices-secure-asp-net-core-mvc-app.aspx)

**Reply by sbwalker** [reply, marked as answer]
@Sharafudeen you are asking a lot of random questions in recent days... AI, blockchain, security best practices, etc... and it is not clear what you are trying to achieve. I would suggest that the best way to learn about Oqtane is to become familiar with Blazor, install Oqtane and review the source code / architecture, and build an Oqtane module yourself. I am including some brief answers to your top 10 security best practices below:

- Cross-Site Scripting (XSS) - Blazor encodes all HTML output by default
- SQL Injection - Oqtane uses EF Core for all data access so there is no chance of SQL injection
- Cross-Site Request Forgery (CSRF) - Oqtane uses anti forgery tokens on all requests to mitigate CSRF
- Custom Error Page for Error Handling - Oqtane has error handling, logging, 404 tracking and redirects
- Version Discloser - is dependent upon whatever web server / environment you use for hosting
- Enforce SSL (Secure Sockets Layer) and HSTS - Oqtane supports HTTPS - you should always use SSL certificates for your sites
- XXE (XML External Entity) Attack - Oqtane does not import XML files
- Improper Authentication and Session Management - Oqtane uses a best of breed BFF security architecture and uses .NET Core Identity for self-hosted auth scenarios
- Sensitive Data Exposure and Audit Trail - Oqtane filters all information disclosed by APIs, and all activities in Oqtane are logged and all entities store created by/on and last modified by/on attributes
- File Upload Validation - Oqtane validates file extensions and file content during upload
